File: src/day14/Solution.java
package day14;

import java.io.IOException;
import java.nio.file.Files;
import java.nio.file.Paths;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.regex.Matcher;
import java.util.regex.Pattern;
import utils.AoCSolvable;
import utils.NoSolutionException;

public class Solution implements AoCSolvable {
  private static final boolean USE_SAMPLE = false;
  private static final Pattern memPattern = Pattern.compile("^mem\\[([0-9]+)] = ([0-9]+)$");
  private final List<String> lines;

  Solution(List<String> lines){
    this.lines = lines;
  }

  public String partOne() throws NoSolutionException {
    Map<Integer, Long> mem = new HashMap<>();
    long maskZero = 0L;
    long maskOne = 0L;
    for (String line: lines){
      if(line.charAt(1)=='a'){
        maskZero = 0L;
        maskOne = 0L;
        String bits = line.substring(7);
        for(char c: bits.toCharArray()){
          maskZero <<= 1;
          maskOne <<= 1;
          if(c == '1'){
            maskOne += 1;
          } else if(c == '0'){
            maskZero += 1;
          }
        }
      } else{
        Matcher m = memPattern.matcher(line);
        m.matches();
        int loc = Integer.parseInt(m.group(1));
        long val = Long.parseLong(m.group(2));
        mem.put(loc, (val | maskOne)& ~maskZero);
      }
    }
    long solution = 0;
    for(int key:mem.keySet()) solution+=mem.get(key);

    return String.valueOf(solution);
  }

  public String partTwo() throws NoSolutionException {
    Map<Long, Long> mem = new HashMap<>();
    long maskOne = 0L;
    long maskX = 0L;
    ArrayList<Long> masks = new ArrayList<>();
    masks.add(0L);
    for (String line: lines){
      if(line.charAt(1)=='a'){
        maskOne = 0L;
        maskX = 0L;
        masks.clear();
        masks.add(0L);
        String bits = line.substring(7);
        for(char c: bits.toCharArray()){
          int maskSize = masks.size();
          maskOne <<= 1;
          maskX <<= 1;
          for(int i = 0; i<maskSize; i++){
            masks.set(i, masks.get(i)<<1);
          }
          if(c == '1'){
            maskOne += 1;
          } else if(c == 'X'){
            maskX+= 1;
            for(int i = 0; i<maskSize; i++){
              masks.add(masks.get(i)+1);
            }
          }
        }
      } else{
        Matcher m = memPattern.matcher(line);
        m.matches();
        long loc = Integer.parseInt(m.group(1));
        long val = Long.parseLong(m.group(2));
        loc = ((loc|maskOne)&~maskX);
        for (long mask:masks){
          long new_loc = loc|mask;
          mem.put(new_loc, val);
        }
      }
    }
    long solution = 0;
    for(long key:mem.keySet()) solution+=mem.get(key);

    return String.valueOf(solution);
  }

  public static void main(String[] args) {
    try{
      List<String> lines = Files.readAllLines(Paths.get((USE_SAMPLE?"./sample/":"./input/")+Solution.class.getPackageName()));
      Solution s = new Solution(lines);
      System.out.println(s.partOne());
      System.out.println(s.partTwo());
    } catch (IOException | NoSolutionException e) {
      e.printStackTrace();
    }
  }
}

File: src/day14/solution.py
import sys
import os
import re
day_number = sys.path[0].split('\\')[-1]
if len(sys.argv)==1:
    path_to_source = os.path.join("\\".join(sys.path[0].split("\\")[:-2]), f"input\\{day_number}")
else:
    path_to_source = os.path.join("\\".join(sys.path[0].split("\\")[:-2]), f"sample\\{day_number}")
with open(path_to_source, "r") as infile:
    lines = infile.read().split("\n")

part1, part2 = None, None
mask_ones = 0 #or
mask_zeros = 0 #nor

memory = dict()
for line in lines:
    
    if(line[:4] == "mask"):
        mask_ones = 0 #and
        mask_zeros = 0 #nor
        bits = line[7:]
        for bit in bits:
            mask_ones <<= 1
            mask_zeros <<= 1
            if(bit == '1'):
                mask_ones += 1
            elif(bit =='0'):
                mask_zeros += 1
    else:
        m = re.match('^mem\[([0-9]+)\] = ([0-9]+)$', line)
        loc = int(m.group(1))
        val = int(m.group(2))
        if(loc not in memory):
            memory[loc] = 0
        memory[loc] = ((val | mask_ones)& ~mask_zeros)
part1 = 0
for loc in memory:
    part1 += memory[loc]
    

memory = dict()
mask_ones = 0
mask_xs = 0
floats = [0]
for line in lines:
    if(line[:4] == "mask"):
        mask_ones = 0
        mask_xs = 0
        floats = [0]
        bits = line[7:]
        for bit in bits:
            mask_ones <<=1
            mask_xs <<=1
            floats = [f<<1 for f in floats]
            if(bit == '1'):
                mask_ones += 1
            elif(bit == 'X'):
                mask_xs += 1
                floats = [f+1 for f in floats]+floats
                
    else:
        m = re.match('^mem\[([0-9]+)\] = ([0-9]+)$', line)
        loc = int(m.group(1))
        val = int(m.group(2))
        loc = ((loc|mask_ones)&~mask_xs)
        for f in floats:
            new_loc = loc | f
            memory[new_loc] = val
part2 = 0
for loc in memory:
    part2 += memory[loc]
print(part1)
print(part2)
        